Fill out all fields for the Instance type

samuelallan72 opened the pull request:
This should completely cover the fields described in the api documentation for [GET /1.0/instances](https://documentation.ubuntu.com/lxd/en/latest/api/#/instances/instance_get).

Note that `restore` is documented there, but does not appear in actual responses (and shouldn't appear; this appears to be a docs issue).  I opened https://github.com/canonical/lxd/issues/12993 to report it.

Fixes #9

Commit:
Fill out all fields for the Instance type

TheSignPainter98 commented on src/client.rs:
```diff
@@ -193,12 +194,14 @@ impl Client {
         Ok(response.json()?)
     }
 
+    /// Get a list of instance names
```
nit: missing full stop ;)

TheSignPainter98 commented on src/instance.rs:
```diff
@@ -40,21 +42,59 @@ pub enum Device {
+
 #[allow(unused)]
 #[derive(Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
```
On second thoughts, I'm not sure whether we want to deny unknown fields here. If lxd gives us an extra field we don't recognise, I think it would be more future-proof to ignore it. Otherwise, I think we'd break our users' code because we haven't gotten around to adding the extra fields!

samuelallan72 replied:
IMHO we definitely want to deny unknown fields.  If lxd updates the api and begins giving an extra field, it would be a breaking change in the same manner as changing the type of a field, and we would need to update the type definition here appropriately.

samuelallan72 replied:
Although if we didn't use deny_unknown_fields, we could then ignore some fields that we didn't care about... hmmm :thinking:

TheSignPainter98 commented on src/instance.rs:
```diff
@@ -40,21 +42,59 @@ pub enum Device {
-    location: String,
-    profiles: Vec<String>,
+    status: InstanceStatus, // TODO: not sure if this string is consistent (could have i18n?)
+    status_code: StatusCode,
```
These two field seem quite similar, it almost looks as though one could call some `instance.status.code()` to get `instance.status_code`!

samuelallan72 replied:
Yeah it's weird, but that's what's returned in the api.   Ideally we would ignore the `status` field and parse the `status_code` into an enum.  Not sure if we can do that neatly with serde though.

samuelallan72 replied:
Yep, I'll merge them.

TheSignPainter98 commented on src/client.rs:
```diff
@@ -107,7 +108,7 @@ struct Response<T> {
     response_type: ResponseType,
     #[allow(unused)]
-    // TODO: enum based on https://documentation.ubuntu.com/lxd/en/latest/rest-api/#list-of-current-status-codes
+    // TODO: enum for http codes - use serde_repr
```
I wonder whether an enum here might be overkill, just looking at `reqwest for inspiration, they use a `struct StatusCode(NonZeroU16)` here, I wonder this simpler route would suffice?

samuelallan72 replied:
I don't think it would be overkill - I was looking at the lxd rest api docs, and they make some strong assertions, like for error return types:

> HTTP code must be one of of 400, 401, 403, 404, 409, 412 or 500.

TheSignPainter98 commented on src/client.rs:
```diff
@@ -193,12 +194,14 @@ impl Client {
             .metadata)
     }
 
+    /// Get an instance's by instance name.
```
nit: I think the wording could be simplified a little :P
```
/// Try to get an instance by its name.
```

TheSignPainter98 commented on src/instance.rs:
```diff
@@ -40,21 +42,59 @@ pub enum Device {
+}
+
+/// Status code from the REST API.
+/// This is not for HTTP status codes.
```
nit: Just to maximise information available to the user, I think we can point them in the right direction here by making the second sentence read something like:
```rust
// For HTTP status codes see [`HttpStatusCode`]
```

Or instead, could the issue be side-stepped entirely by renaming this to `InstanceStatusCode`?

samuelallan72 replied:
Good idea, I've renamed it. :)

samuelallan72 replied:
Agreed, I've renamed it. :)

TheSignPainter98 commented on src/instance.rs:
```diff
@@ -40,21 +42,59 @@ pub enum Device {
-// TODO: deny unknown fields once Instance type is complete
-// #[serde(deny_unknown_fields)]
+#[derive(Debug, Deserialize)]
+pub enum InstanceStatus {
```
Should these variants include the status code? If every `StatusCode` appears alongside at most one `InstanceStatus`, perhaps we could encode these together? As in could we have something like this?
```rust
pub enum InstanceStatus {
    Running(RunningStatus),
    ...
}

pub enum RunningStatus {
    OperationCreated = 100,
    ...
}
```
I have a feeling some combinations of `(instance.status, instance.status_code)` may be impossible, e.g. `(InstanceStatus::Error, StatusCode::Success)`

samuelallan72 replied:
It appears that `status` is simple the string representation of `status_code`.  https://github.com/canonical/lxd/blob/main/shared/api/status_code.go

So the two fields are redundant.  I'll see how we can ignore the `status` field I think.

TheSignPainter98 commented on src/instance.rs:
```diff
@@ -40,21 +42,59 @@ pub enum Device {
+#[serde(deny_unknown_fields)]
 pub struct Instance {
+    architecture: String,
+    config: HashMap<String, String>, // TODO: is config completely described - can we use a struct?
```
I think merging TODOs can be okay, but I'd prefer that we didn't merge a TODO with a question mark over it :P

TheSignPainter98 commented on src/instance.rs:
```diff
@@ -40,21 +42,59 @@ pub enum Device {
+    devices: HashMap<String, Device>,
+    ephemeral: bool,
+    expanded_config: HashMap<String, String>,
+    expanded_devices: HashMap<String, Device>,
```
I'm not quite sure I follow, do we need to keep around both `devices` and `expanded_devices`? (Same for `config`)

samuelallan72 replied:
I'm not sure.  It seems that the two could be different, and perhaps useful to know.  The docs say:

expanded_devices:
> Expanded devices (all profiles and local devices merged)

devices:
> Instance devices (see doc/instances.md)

So I guess one is the final consolidated list of devices given the profiles, etc., while the other is just the devices configured specifically for that instance.

TheSignPainter98 reviewed (changes requested):
This is coming along nicely! Here are a few things to consider, mostly nitpicks though

Commit:
Merge status and status_code

Since they are redundant.

Also improve some docs.

samuelallan72 commented:
Thanks @TheSignPainter98 , I fixed the docs where noted, and and merged status and status_code into a more complete instance status.  I also removed `deny_unknown_fields` for `Instance` because otherwise it would be messy to ignore the redundant status field.  I think I would prefer to retain `deny_unknown_fields` to ensure completeness, but I guess in the end it doesn't matter too much?

samuelallan72 commented:
moved to https://github.com/canonical/nlxd-rs/pull/14
